tests/unit/utils/test_collections.py

```diff
@@ -0,0 +1,49 @@
+from dvc.utils.collections import apply_diff
+
+
+class MyDict(dict):
+    pass
+
+
+class MyInt(int):
+    pass
+
+
+def test_apply_diff_is_inplace():
+    dest = MyDict()
+    dest.attr = 42
+    apply_diff({}, dest)
+
+    assert type(dest) is MyDict, "Preserves class"
+    assert dest.attr == 42, "Preserves custom attrs"
+
+
+def test_apply_diff_mapping():
+    src = {"a": 1}
+    dest = {"b": 2}
+    apply_diff(src, dest)
+    assert dest == src, "Adds and removes keys"
+
+    src = {"a": 1}
+    dest = {"a": MyInt(1)}
+    apply_diff(src, dest)
+    assert type(dest["a"]) is MyInt, "Does not replace equals"
+
+    src = {"d": {"a": 1}}
+    inner = {}
+    dest = {"d": inner}
+    apply_diff(src, dest)
+    assert dest["d"] is inner, "Updates inner dicts"
+
+
```

Comment:
Still not testing that ValueError that you are raising in apply_diff. :slightly_smiling_face:

Reply:
This test would have no value. This is not target behavior to test, this will be simple code duplication.

Reply:
We don't even know if that code would work without at least unit testing it. We should either test it or remove it and/or replace it with an assert.

Reply:
This is not code implementing any desirable behavior, we don't really care what it does as long as it's loud. And I am pretty sure that raise will work.

Reply:
Say I add code to test for `ValueError`, later I will change it to `TypeError` and test would be needed to change the same. A need for coordinated changes is a symptom of code duplication.

Reply:
Please correct me if I'm wrong, but as I understand, this code is a "just in case" one and usually, it shouldn't execute at all within dvc. If it does run -- then we have a problem somewhere, since it is not a part of normal operation. Nothing screams more loudly `BUG!!!` than an assert exception. I would agree that this `ValueError` is needed if it was a public method in our package and it was meant to be used by users, but if that was the case, it _should_ be tested with no doubts. So we should either test it with a unit test, or we should use an assert there.

Reply:
If you pass dict to `Stage.load()` as `fname` that will give you `ValueError` or `TypeError` are you suggesting unit testing it? Or adding an assert in every function checking all argument types?

Reply:
@Suor If `Stage.load()` had that logic for checking `fname` type and would raise Value/TypeError explicitly then yes, I would add a unit test for it to ensure that it works as expected.

I'm saying that if you are leaving ValueError in `apply_diff` and you are unit testing `apply_diff`, you should test for that `ValueError` as well, or replace it with an `assert` and not test it at all.

Reply:
I changed this to assert, time will tell who was right)

Reply:
@Suor Thanks! :slightly_smiling_face: But now your assert doesn't match your `ValueError` that you've had before and has additional logic in it. It used to be:
```
if isinstance(src, Mapping) and isinstance(dest, Mapping):
    ...
elif isinstance(src, Seq) and isinstance(dest, Seq):
    ...
else:
    raise ValueError()
```

and I meant that you could use something like
```
if isinstance(src, Mapping) and isinstance(dest, Mapping):
    ...
elif isinstance(src, Seq) and isinstance(dest, Seq):
    ...
else:
    assert False
```

if you don't wish to test it. In this form, this is a simple defensive `assert` and there is no additional logic in the assert itself. And now you have
```
assert is_same_type(src, dst)
if isinstance(src, Mapping) and isinstance(dest, Mapping):
    ...
elif isinstance(src, Seq) and isinstance(dest, Seq):
    ...
```
which is not equivalent to your initial logic, right? And now your assert has additional logic in it, but at least it is not dead code and is actually run during normal operation.

Reply:
My asserts do much. Yours don't test for src and dest being both strings or bools.

Reply:
New version of the code is still worse. There should be `else` and raise `ValueError`, this will guarantee that condition match is exhaustive and error type would make sense.

Reply:
Your initial `ValueError` didn't test for that either and I was just proposing to use assert instead if you are so unwilling to test your `ValueError`.